
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="X-UA-Compatible" content="IE=Edge" />
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>geppy.core.entity &#8212; geppy 0.1.0a0 documentation</title>
    <link rel="stylesheet" href="../../../_static/pydoctheme.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <script type="text/javascript" id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
    <script type="text/javascript" src="../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
 
<script type="text/javascript">
  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-15144370-3']);
  _gaq.push(['_trackPageview']);
</script>
<script type="text/javascript" src="../../../_static/copybutton.js"></script>

  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
    <li><a href="https://github.com/ShuhuaGao/geppy"><img src="../../../_static/geppy-icon.png" style="height: 30px; vertical-align: middle; padding-right: 1em;" />Project Homepage</a> &#187;</li>
	<li><a href="../../../index.html">geppy 0.1.0a0 documentation</a> &#187; </li>

          <li class="nav-item nav-item-1"><a href="../../index.html" accesskey="U">Module code</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for geppy.core.entity</h1><div class="highlight"><pre>
<span></span><span class="c1"># coding=utf-8</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="sd">.. moduleauthor:: Shuhua Gao</span>

<span class="sd">The module :mod:`entity` defines the core data structures used in GEP, including the gene, the chromosome, the</span>
<span class="sd">K-expression and the expression tree. We use primitives (functions and terminals) to build genes and then compose</span>
<span class="sd">a chromosome with one or multiple genes. Refer to the :mod:`geppy.core.symbol` module for information on primitives.</span>

<span class="sd">A chromosome composed of only one gene is called a *monogenic* chromosome, while one composed of more than one genes</span>
<span class="sd">is named a *multigenic* chromosome. A multigenic chromosome can be assigned a linking function to combine the results</span>
<span class="sd">from multiple genes into a single result.</span>
<span class="sd">&quot;&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">copy</span>
<span class="kn">from</span> <span class="nn">..tools.generator</span> <span class="k">import</span> <span class="o">*</span>
<span class="kn">from</span> <span class="nn">..core.symbol</span> <span class="k">import</span> <span class="n">Function</span><span class="p">,</span> <span class="n">RNCTerminal</span><span class="p">,</span> <span class="n">Terminal</span>


<span class="n">_DEBUG</span> <span class="o">=</span> <span class="kc">False</span>


<div class="viewcode-block" id="KExpression"><a class="viewcode-back" href="../../../geppy.core.html#geppy.core.entity.KExpression">[docs]</a><span class="k">class</span> <span class="nc">KExpression</span><span class="p">(</span><span class="nb">list</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Class representing the K-expression, or the open reading frame (ORF), of a gene in GEP. A K-expression is actually</span>
<span class="sd">    a linear form of an expression tree obtained by level-order traversal.</span>
<span class="sd">    &quot;&quot;&quot;</span>
<div class="viewcode-block" id="KExpression.__init__"><a class="viewcode-back" href="../../../geppy.core.html#geppy.core.entity.KExpression.__init__">[docs]</a>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">content</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize a K-expression.</span>

<span class="sd">        :param content: iterable, each element is a primitive (function or terminal)</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">list</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">content</span><span class="p">)</span></div>

<div class="viewcode-block" id="KExpression.__str__"><a class="viewcode-back" href="../../../geppy.core.html#geppy.core.entity.KExpression.__str__">[docs]</a>    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get a string representation of this expression by joining the name of each primitive.</span>
<span class="sd">        :return:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="s1">&#39;[&#39;</span> <span class="o">+</span> <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">ele</span><span class="o">.</span><span class="n">name</span> <span class="k">for</span> <span class="n">ele</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;]&#39;</span></div>

    <span class="k">def</span> <span class="nf">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;[</span><span class="si">{}</span><span class="s1">]&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">repr</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">))</span></div>


<div class="viewcode-block" id="ExpressionTree"><a class="viewcode-back" href="../../../geppy.core.html#geppy.core.entity.ExpressionTree">[docs]</a><span class="k">class</span> <span class="nc">ExpressionTree</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Class representing an expression tree (ET) in GEP, which may be obtained by translating a K-expression, a</span>
<span class="sd">    gene, or a chromosome, i.e., genotype-phenotype mapping.</span>
<span class="sd">    &quot;&quot;&quot;</span>
<div class="viewcode-block" id="ExpressionTree.__init__"><a class="viewcode-back" href="../../../geppy.core.html#geppy.core.entity.ExpressionTree.__init__">[docs]</a>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">root</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize a tree with the given *root* node.</span>

<span class="sd">        :param root: :class:`ExpressionTree.Node`, the root node</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_root</span> <span class="o">=</span> <span class="n">root</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">root</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the root node of this expression tree.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_root</span>

<div class="viewcode-block" id="ExpressionTree.Node"><a class="viewcode-back" href="../../../geppy.core.html#geppy.core.entity.ExpressionTree.Node">[docs]</a>    <span class="k">class</span> <span class="nc">Node</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Class representing a node in the expression tree. Each node has a variable number of children, depending on</span>
<span class="sd">        the arity of the primitive at this node.</span>
<span class="sd">        &quot;&quot;&quot;</span>
<div class="viewcode-block" id="ExpressionTree.Node.__init__"><a class="viewcode-back" href="../../../geppy.core.html#geppy.core.entity.ExpressionTree.Node.__init__">[docs]</a>        <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">):</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_children</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_name</span> <span class="o">=</span> <span class="n">name</span></div>

        <span class="nd">@property</span>
        <span class="k">def</span> <span class="nf">children</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
            <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            Get the children of this node.</span>
<span class="sd">            &quot;&quot;&quot;</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_children</span>

        <span class="nd">@property</span>
        <span class="k">def</span> <span class="nf">name</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
            <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">            Get the name (label) of this node.</span>
<span class="sd">            &quot;&quot;&quot;</span>
            <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_name</span></div>

<div class="viewcode-block" id="ExpressionTree.from_genotype"><a class="viewcode-back" href="../../../geppy.core.html#geppy.core.entity.ExpressionTree.from_genotype">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">from_genotype</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">genome</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create an expression tree by translating *genome*, which may be a K-expression, a gene, or a chromosome.</span>

<span class="sd">        :param genome: :class:`KExpression`, :class:`Gene`, or :class:`Chromosome`, the genotype of an individual</span>
<span class="sd">        :return: :class:`ExpressionTree`, an expression tree</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">Gene</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_from_kexpression</span><span class="p">(</span><span class="n">genome</span><span class="o">.</span><span class="n">kexpression</span><span class="p">)</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">KExpression</span><span class="p">):</span>
            <span class="k">return</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_from_kexpression</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span>
        <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">Chromosome</span><span class="p">):</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">genome</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                <span class="k">return</span> <span class="bp">cls</span><span class="o">.</span><span class="n">_from_kexpression</span><span class="p">(</span><span class="n">genome</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">kexpression</span><span class="p">)</span>
            <span class="n">sub_trees</span> <span class="o">=</span> <span class="p">[</span><span class="bp">cls</span><span class="o">.</span><span class="n">_from_kexpression</span><span class="p">(</span><span class="n">gene</span><span class="o">.</span><span class="n">kexpression</span><span class="p">)</span> <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">genome</span><span class="p">]</span>
            <span class="c1"># combine the sub_trees with the linking function</span>
            <span class="n">root</span> <span class="o">=</span> <span class="bp">cls</span><span class="o">.</span><span class="n">Node</span><span class="p">(</span><span class="n">genome</span><span class="o">.</span><span class="n">linker</span><span class="o">.</span><span class="vm">__name__</span><span class="p">)</span>
            <span class="n">root</span><span class="o">.</span><span class="n">children</span><span class="p">[:]</span> <span class="o">=</span> <span class="n">sub_trees</span>
            <span class="k">return</span> <span class="bp">cls</span><span class="p">(</span><span class="n">root</span><span class="p">)</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="s1">&#39;Only an argument of type KExpression, Gene, and Chromosome is acceptable. The provided &#39;</span>
                        <span class="s1">&#39;genome type is </span><span class="si">{}</span><span class="s1">.&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">genome</span><span class="p">)))</span></div>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">_from_kexpression</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">expr</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create an expression tree from a K-expression.</span>

<span class="sd">        :param expr: a K-expression</span>
<span class="sd">        :return: :class:`ExpressionTree`, an expression tree</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">expr</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="c1"># first build a node for each primitive</span>
        <span class="n">nodes</span> <span class="o">=</span> <span class="p">[</span><span class="bp">cls</span><span class="o">.</span><span class="n">Node</span><span class="p">(</span><span class="n">p</span><span class="o">.</span><span class="n">name</span><span class="p">)</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">expr</span><span class="p">]</span>
        <span class="c1"># connect each node to its children if any</span>
        <span class="n">i</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">j</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">while</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">nodes</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">expr</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">arity</span><span class="p">):</span>
                <span class="n">j</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">nodes</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">children</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">nodes</span><span class="p">[</span><span class="n">j</span><span class="p">])</span>
            <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">return</span> <span class="bp">cls</span><span class="p">(</span><span class="n">nodes</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span></div>


<div class="viewcode-block" id="Gene"><a class="viewcode-back" href="../../../geppy.core.html#geppy.core.entity.Gene">[docs]</a><span class="k">class</span> <span class="nc">Gene</span><span class="p">(</span><span class="nb">list</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A single gene in GEP, which is a fixed-length linear sequence composed of functions and terminals.</span>
<span class="sd">    &quot;&quot;&quot;</span>
<div class="viewcode-block" id="Gene.__init__"><a class="viewcode-back" href="../../../geppy.core.html#geppy.core.entity.Gene.__init__">[docs]</a>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pset</span><span class="p">,</span> <span class="n">head_length</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Instantiate a gene.</span>
<span class="sd">        :param head_length: length of the head domain</span>
<span class="sd">        :param pset: a primitive set including functions and terminals for genome construction.</span>

<span class="sd">        Supposing the maximum arity of functions in *pset* is *max_arity*, then the tail length is automatically</span>
<span class="sd">        determined to be ``tail_length = head_length * (max_arity - 1) + 1``. The genome, i.e., list of symbols in</span>
<span class="sd">        the instantiated gene is formed randomly from *pset*.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_head_length</span> <span class="o">=</span> <span class="n">head_length</span>
        <span class="n">genome</span> <span class="o">=</span> <span class="n">generate_genome</span><span class="p">(</span><span class="n">pset</span><span class="p">,</span> <span class="n">head_length</span><span class="p">)</span>
        <span class="nb">list</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genome</span><span class="p">)</span></div>

<div class="viewcode-block" id="Gene.from_genome"><a class="viewcode-back" href="../../../geppy.core.html#geppy.core.entity.Gene.from_genome">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">from_genome</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">head_length</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Build a gene directly from the given *genome*.</span>

<span class="sd">        :param genome: iterable, a list of symbols representing functions and terminals</span>
<span class="sd">        :param head_length: length of the head domain</span>
<span class="sd">        :return: :class:`Gene`, a gene</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">g</span> <span class="o">=</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__new__</span><span class="p">(</span><span class="bp">cls</span><span class="p">)</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="n">genome</span><span class="p">)</span>
        <span class="n">g</span><span class="o">.</span><span class="n">_head_length</span> <span class="o">=</span> <span class="n">head_length</span>
        <span class="k">return</span> <span class="n">g</span></div>

<div class="viewcode-block" id="Gene.__str__"><a class="viewcode-back" href="../../../geppy.core.html#geppy.core.entity.Gene.__str__">[docs]</a>    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return the expression in a human readable string, which is also a legal python code that can be evaluated.</span>

<span class="sd">        :return: string form of the expression</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">expr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">kexpression</span>
        <span class="n">i</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">expr</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="k">while</span> <span class="n">i</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">expr</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">arity</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>  <span class="c1"># a function</span>
                <span class="n">f</span> <span class="o">=</span> <span class="n">expr</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                <span class="n">args</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">arity</span><span class="p">):</span>
                    <span class="n">ele</span> <span class="o">=</span> <span class="n">expr</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
                    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ele</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                        <span class="n">args</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ele</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>  <span class="c1"># a terminal or an ephemeral class</span>
                        <span class="n">args</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ele</span><span class="o">.</span><span class="n">format</span><span class="p">())</span>
                <span class="n">expr</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="o">*</span><span class="nb">reversed</span><span class="p">(</span><span class="n">args</span><span class="p">))</span>  <span class="c1"># replace the operator with its result (str)</span>
            <span class="n">i</span> <span class="o">-=</span> <span class="mi">1</span>

        <span class="c1"># the final result is at the root. It may happen that the K-expression contains only one terminal at its root.</span>
        <span class="k">return</span> <span class="n">expr</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">expr</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">str</span><span class="p">)</span> <span class="k">else</span> <span class="n">expr</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">format</span><span class="p">()</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">kexpression</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;Get the K-expression of type :class:`KExpression` represented by this gene.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># get the level-order K expression</span>
        <span class="n">expr</span> <span class="o">=</span> <span class="n">KExpression</span><span class="p">([</span><span class="bp">self</span><span class="p">[</span><span class="mi">0</span><span class="p">]])</span>
        <span class="n">i</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">j</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">while</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">expr</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">expr</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">arity</span><span class="p">):</span>
                <span class="n">expr</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="bp">self</span><span class="p">[</span><span class="n">j</span><span class="p">])</span>
                <span class="n">j</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">return</span> <span class="n">expr</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">head_length</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the length of the head domain.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_head_length</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">tail_length</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the length of the tail domain.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">head_length</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">max_arity</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the max arity of the functions in the primitive set with which the gene is built.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">//</span> <span class="bp">self</span><span class="o">.</span><span class="n">head_length</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">head</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the head domain of the gene.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">head_length</span><span class="p">]</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">tail</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the tail domain of the gene.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">head_length</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">head_length</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">tail_length</span><span class="p">]</span>

    <span class="k">def</span> <span class="nf">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return a string representation of a gene including all the symbols. Mainly for debugging purpose.</span>

<span class="sd">        :return: a string</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">info</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">p</span><span class="p">)</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">]</span>
        <span class="k">return</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1"> [&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">info</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;]&#39;</span></div>


<div class="viewcode-block" id="GeneDc"><a class="viewcode-back" href="../../../geppy.core.html#geppy.core.entity.GeneDc">[docs]</a><span class="k">class</span> <span class="nc">GeneDc</span><span class="p">(</span><span class="n">Gene</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Class represents a gene with an additional Dc domain to handle numerical constants in GEP.</span>

<span class="sd">    The basic :class:`Gene`</span>
<span class="sd">    has two domains, a head and a tail, while this :class:`GeneDc` class introduces another domain called *Dc*</span>
<span class="sd">    after the tail. The length of the Dc domain is equal to the length of the tail domain. The Dc domain only stores</span>
<span class="sd">    the indices of numbers present in a separate array :meth:`rnc_array`, which collects a group of candidate</span>
<span class="sd">    random numerical constants (RNCs). Thus, each :class:`GeneDc` instance comes with a *rnc_array*, which are</span>
<span class="sd">    generally different among different instances.</span>

<span class="sd">    In addition to the operators of the basic gene expression algorithm (mutation, inversion, transposition, and</span>
<span class="sd">    recombination), Dc-specific operators in GEP-RNC are also created to better evolve the Dc domain and to manipulate</span>
<span class="sd">    the attached set of constants *rnc_array*. Such operators are all suffixed with &#39;_dc&#39; in</span>
<span class="sd">    :mod:`~geppy.tools.crossover` and :mod:`~geppy.tools.mutation` modules, for example, the method</span>
<span class="sd">    :func:`~geppy.tools.mutation.invert_dc`.</span>

<span class="sd">    The *rnc_array* associated with each :class:`GeneDc` instance can be provided explicitly when creating</span>
<span class="sd">    a :class:`GeneDc` instance. See :meth:`from_genome`. Or more generally, a</span>
<span class="sd">    random number generator *rnc_gen* can be provided, with which a specified number of RNCs are generated during the</span>
<span class="sd">    creation of gene instances.</span>

<span class="sd">    A special terminal of type :class:`~geppy.core.symbol.TerminalRNC` is used internally to represent the RNCs.</span>

<span class="sd">    .. note::</span>
<span class="sd">        To create an effective :class:`GeneDc` gene, the primitive set should contain at least one</span>
<span class="sd">        :class:`~geppy.core.symbol.TerminalRNC` terminal. See :meth:`~geppy.core.symbol.PrimitiveSet.add_rnc`</span>
<span class="sd">        for details.</span>

<span class="sd">    Refer to Chapter 5 of [FC2006]_ for more knowledge about GEP-RNC.</span>
<span class="sd">    &quot;&quot;&quot;</span>
<div class="viewcode-block" id="GeneDc.__init__"><a class="viewcode-back" href="../../../geppy.core.html#geppy.core.entity.GeneDc.__init__">[docs]</a>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">pset</span><span class="p">,</span> <span class="n">head_length</span><span class="p">,</span> <span class="n">rnc_gen</span><span class="p">,</span> <span class="n">rnc_array_length</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize a gene with a Dc domain.</span>

<span class="sd">        :param head_length: length of the head domain</span>
<span class="sd">        :param pset: a primitive set including functions and terminals for genome construction.</span>
<span class="sd">        :param rnc_gen: callable, which should generate a random number when called by ``rnc_gen()``.</span>
<span class="sd">        :param rnc_array_length: int, number of random numerical constant candidates associated with this gene,</span>
<span class="sd">            usually 10 is enough</span>

<span class="sd">        Supposing the maximum arity of functions in *pset* is *max_arity*, then the tail length is automatically</span>
<span class="sd">        determined to be ``tail_length = head_length * (max_arity - 1) + 1``. The genome, i.e., list of symbols in</span>
<span class="sd">        the instantiated gene is formed randomly from *pset*. The length of Dc domain is equal to *tail_length*, i.e.,</span>
<span class="sd">        the tail domain and the Dc domain share the same length.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># first generate the gene without Dc</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">pset</span><span class="p">,</span> <span class="n">head_length</span><span class="p">)</span>
        <span class="n">t</span> <span class="o">=</span> <span class="n">head_length</span> <span class="o">*</span> <span class="p">(</span><span class="n">pset</span><span class="o">.</span><span class="n">max_arity</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="n">d</span> <span class="o">=</span> <span class="n">t</span>
        <span class="c1"># complement it with a Dc domain</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_rnc_gen</span> <span class="o">=</span> <span class="n">rnc_gen</span>
        <span class="n">dc</span> <span class="o">=</span> <span class="n">generate_dc</span><span class="p">(</span><span class="n">rnc_array_length</span><span class="p">,</span> <span class="n">dc_length</span><span class="o">=</span><span class="n">d</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">dc</span><span class="p">)</span>
        <span class="c1"># generate the rnc array</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_rnc_array</span> <span class="o">=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">_rnc_gen</span><span class="p">()</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">rnc_array_length</span><span class="p">)]</span></div>

<div class="viewcode-block" id="GeneDc.from_genome"><a class="viewcode-back" href="../../../geppy.core.html#geppy.core.entity.GeneDc.from_genome">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">from_genome</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">genome</span><span class="p">,</span> <span class="n">head_length</span><span class="p">,</span> <span class="n">rnc_array</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Build a gene directly from the given *genome* and the random numerical constant (RNC) array *rnc_array*.</span>

<span class="sd">        :param genome: iterable, a list of symbols representing functions and terminals (especially the special RNC</span>
<span class="sd">            terminal of type :class:`~geppy.core.symbol.TerminalRNC`). This genome should have three</span>
<span class="sd">            domains: head, tail and Dc. The Dc domain should be composed only of integers representing indices into the</span>
<span class="sd">            *rnc_array*.</span>
<span class="sd">        :param head_length: length of the head domain. The length of the tail and Dc domain should follow the rule and</span>
<span class="sd">            can be determined from the head_length: ``dc_length = tail_length = (len(genome) - head_length) / 2``.</span>
<span class="sd">        :param rnc_array: the RNC array associated with the gene, which contains random constant candidates</span>
<span class="sd">        :return: :class:`GeneDc`, a gene</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">g</span> <span class="o">=</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="n">from_genome</span><span class="p">(</span><span class="n">genome</span><span class="p">,</span> <span class="n">head_length</span><span class="p">)</span>  <span class="c1"># the genome is copied and the head-length is fixed</span>
        <span class="n">g</span><span class="o">.</span><span class="n">_rnc_array</span> <span class="o">=</span> <span class="n">copy</span><span class="o">.</span><span class="n">deepcopy</span><span class="p">(</span><span class="n">rnc_array</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">g</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">dc_length</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the length of the Dc domain, which is equal to :meth:`GeneDc.tail_length`</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">tail_length</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">rnc_array</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the random numerical array (RNC) associated with this gene.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_rnc_array</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">tail_length</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the tail domain length.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-</span> <span class="bp">self</span><span class="o">.</span><span class="n">head_length</span><span class="p">)</span> <span class="o">//</span> <span class="mi">2</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">max_arity</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the max arity of functions in the primitive set used to build this gene.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># determine from the head-tail-Dc length relations</span>
        <span class="k">return</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-</span> <span class="mi">2</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">head_length</span><span class="p">)</span> <span class="o">//</span> <span class="p">(</span><span class="mi">2</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">head_length</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">dc</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the Dc domain of this gene.</span>
<span class="sd">        :return:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">head_length</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">tail_length</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">head_length</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">tail_length</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">dc_length</span><span class="p">]</span>

    <span class="c1"># def __deepcopy__(self, memodict):</span>
    <span class="c1">#     return self.__class__.from_genome(self, self.head_length, self.rnc_array)</span>

<div class="viewcode-block" id="GeneDc.__str__"><a class="viewcode-back" href="../../../geppy.core.html#geppy.core.entity.GeneDc.__str__">[docs]</a>    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return the expression in a human readable string, which is also a legal python code that can be evaluated.</span>
<span class="sd">        The special terminal representing a RNC will be replaced by their true values retrieved from the array</span>
<span class="sd">        :meth:`rnc_array`.</span>

<span class="sd">        :return: string form of the expression</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">expr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">kexpression</span>
        <span class="n">n_total_rncs</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="mi">1</span> <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">RNCTerminal</span><span class="p">)</span> <span class="k">else</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">expr</span><span class="p">)</span>  <span class="c1"># how many RNCs in total?</span>
        <span class="n">n_encountered_rncs</span> <span class="o">=</span> <span class="mi">0</span>  <span class="c1"># how many RNCs we have encountered in reverse order?</span>
        <span class="n">i</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">expr</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>
        <span class="k">while</span> <span class="n">i</span> <span class="o">&gt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">expr</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">arity</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>  <span class="c1"># a function</span>
                <span class="n">f</span> <span class="o">=</span> <span class="n">expr</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                <span class="n">args</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">arity</span><span class="p">):</span>
                    <span class="n">ele</span> <span class="o">=</span> <span class="n">expr</span><span class="o">.</span><span class="n">pop</span><span class="p">()</span>
                    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ele</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
                        <span class="n">args</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ele</span><span class="p">)</span>
                    <span class="k">else</span><span class="p">:</span>  <span class="c1"># a terminal or a RNC terminal</span>
                        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ele</span><span class="p">,</span> <span class="n">RNCTerminal</span><span class="p">):</span>
                            <span class="c1"># retrieve its true value</span>
                            <span class="n">which</span> <span class="o">=</span> <span class="n">n_total_rncs</span> <span class="o">-</span> <span class="n">n_encountered_rncs</span> <span class="o">-</span> <span class="mi">1</span>
                            <span class="n">index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dc</span><span class="p">[</span><span class="n">which</span><span class="p">]</span>
                            <span class="n">value</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">rnc_array</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>
                            <span class="n">args</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">value</span><span class="p">))</span>
                            <span class="n">n_encountered_rncs</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="k">else</span><span class="p">:</span>   <span class="c1"># a normal terminal</span>
                            <span class="n">args</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">ele</span><span class="o">.</span><span class="n">format</span><span class="p">())</span>
                <span class="n">expr</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="o">*</span><span class="nb">reversed</span><span class="p">(</span><span class="n">args</span><span class="p">))</span>  <span class="c1"># replace the operator with its result (str)</span>
            <span class="n">i</span> <span class="o">-=</span> <span class="mi">1</span>

        <span class="c1"># the final result is at the root</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">expr</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="nb">str</span><span class="p">):</span>
            <span class="k">return</span> <span class="n">expr</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">expr</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">RNCTerminal</span><span class="p">):</span>  <span class="c1"># only contains a single RNC, let&#39;s retrieve its value</span>
            <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">rnc_array</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">dc</span><span class="p">[</span><span class="mi">0</span><span class="p">]])</span>
        <span class="k">return</span> <span class="n">expr</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">format</span><span class="p">()</span>    <span class="c1"># only contains a normal terminal</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">kexpression</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the K-expression of type :class:`KExpression` represented by this gene. The involved RNC terminal will be</span>
<span class="sd">        replaced by a constant terminal with its value retrived from the :meth:`rnc_array` according to the GEP-RNC</span>
<span class="sd">        algorithm.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">n_rnc</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">def</span> <span class="nf">convert_RNC</span><span class="p">(</span><span class="n">p</span><span class="p">):</span>
            <span class="k">nonlocal</span> <span class="n">n_rnc</span>
            <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="n">RNCTerminal</span><span class="p">):</span>
                <span class="n">index</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">dc</span><span class="p">[</span><span class="n">n_rnc</span><span class="p">]</span>
                <span class="n">value</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">rnc_array</span><span class="p">[</span><span class="n">index</span><span class="p">]</span>
                <span class="n">n_rnc</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">t</span> <span class="o">=</span> <span class="n">Terminal</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">value</span><span class="p">),</span> <span class="n">value</span><span class="p">)</span>
                <span class="k">return</span> <span class="n">t</span>
            <span class="k">return</span> <span class="n">p</span>

        <span class="c1"># level-order</span>
        <span class="n">expr</span> <span class="o">=</span> <span class="n">KExpression</span><span class="p">([</span><span class="n">convert_RNC</span><span class="p">(</span><span class="bp">self</span><span class="p">[</span><span class="mi">0</span><span class="p">])])</span>
        <span class="n">i</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">j</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="k">while</span> <span class="n">i</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">expr</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">expr</span><span class="p">[</span><span class="n">i</span><span class="p">]</span><span class="o">.</span><span class="n">arity</span><span class="p">):</span>
                <span class="n">expr</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">convert_RNC</span><span class="p">(</span><span class="bp">self</span><span class="p">[</span><span class="n">j</span><span class="p">]))</span>
                <span class="n">j</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="n">i</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">return</span> <span class="n">expr</span>

    <span class="k">def</span> <span class="nf">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__repr__</span><span class="p">()</span> <span class="o">+</span> <span class="s1">&#39;, rnc_array=[&#39;</span> <span class="o">+</span> <span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">num</span><span class="p">)</span> <span class="k">for</span> <span class="n">num</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">rnc_array</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;]&#39;</span></div>


<div class="viewcode-block" id="Chromosome"><a class="viewcode-back" href="../../../geppy.core.html#geppy.core.entity.Chromosome">[docs]</a><span class="k">class</span> <span class="nc">Chromosome</span><span class="p">(</span><span class="nb">list</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Individual representation in gene expression programming (GEP), which may contain one or more genes. Note that in a</span>
<span class="sd">    multigenic chromosome, all genes must share the same head length and the same primitive set. Each element of</span>
<span class="sd">    :class:`Chromosome` is of type :class:`Gene`.</span>
<span class="sd">    &quot;&quot;&quot;</span>

<div class="viewcode-block" id="Chromosome.__init__"><a class="viewcode-back" href="../../../geppy.core.html#geppy.core.entity.Chromosome.__init__">[docs]</a>    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">gene_gen</span><span class="p">,</span> <span class="n">n_genes</span><span class="p">,</span> <span class="n">linker</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize an individual in GEP.</span>

<span class="sd">        :param gene_gen: callable, a gene generator, i.e., ``gene_gen()`` should return a gene</span>
<span class="sd">        :param n_genes: number of genes</span>
<span class="sd">        :param linker: callable, a linking function</span>

<span class="sd">        .. note:</span>
<span class="sd">            If a linker is specified, then it must accept *n_genes* arguments, each produced by one gene. If the</span>
<span class="sd">            *linker* parameter is the default value ``None``, then for a monogenic chromosome, no linking is applied,</span>
<span class="sd">            while for a multigenic chromosome, the ``tuple`` function is used.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="nb">list</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="p">(</span><span class="n">gene_gen</span><span class="p">()</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_genes</span><span class="p">)))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_linker</span> <span class="o">=</span> <span class="n">linker</span></div>

<div class="viewcode-block" id="Chromosome.from_genes"><a class="viewcode-back" href="../../../geppy.core.html#geppy.core.entity.Chromosome.from_genes">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">from_genes</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">genes</span><span class="p">,</span> <span class="n">linker</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create a chromosome instance by providing a list of genes directly. The number of genes is implicitly specified</span>
<span class="sd">        by ``len(genes)``.</span>

<span class="sd">        :param genes: iterable, a list of genes</span>
<span class="sd">        :param linker: callable, a linking function. Refer to :meth:`Chromosome.__init__` for more details.</span>
<span class="sd">        :return: :class:`Chromosome`, a chromosome</span>

<span class="sd">        .. caution:</span>
<span class="sd">            Every gene of type :class:`Gene` in *genes* should be an independent object. That is, for any two genes</span>
<span class="sd">            *g1* and *g2*, the assertion ``g1 is g1`` should return ``False``. Otherwise, unexpected results may happen</span>
<span class="sd">            in following evolution.</span>

<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">c</span> <span class="o">=</span> <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__new__</span><span class="p">(</span><span class="bp">cls</span><span class="p">)</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="n">genes</span><span class="p">)</span>
        <span class="n">c</span><span class="o">.</span><span class="n">_linker</span> <span class="o">=</span> <span class="n">linker</span>
        <span class="k">return</span> <span class="n">c</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">head_length</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the length of the head domain. All genes in a chromosome should have the same head length.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">head_length</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">tail_length</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the length of the tail domain. All genes in a chromosome should have the same tail length.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">tail_length</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">max_arity</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the max arity of the functions in the primitive set with which all genes are built. Note that all genes</span>
<span class="sd">        in a chromosome should share the same primitive set.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="bp">self</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">max_arity</span>

<div class="viewcode-block" id="Chromosome.__str__"><a class="viewcode-back" href="../../../geppy.core.html#geppy.core.entity.Chromosome.__str__">[docs]</a>    <span class="k">def</span> <span class="nf">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return the expressions in a human readable string.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">linker</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="s1">&#39;</span><span class="si">{0}</span><span class="s1">(</span><span class="se">\n\t</span><span class="si">{1}</span><span class="se">\n</span><span class="s1">)&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">linker</span><span class="o">.</span><span class="vm">__name__</span><span class="p">,</span> <span class="s1">&#39;,</span><span class="se">\n\t</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">g</span><span class="p">)</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">assert</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span>
            <span class="k">return</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">kexpressions</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get a list of K-expressions for all the genes in this chromosome.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">[</span><span class="n">g</span><span class="o">.</span><span class="n">kexpression</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">]</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">linker</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Get the linking function.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_linker</span> <span class="ow">is</span> <span class="kc">None</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="k">return</span> <span class="nb">tuple</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_linker</span>

    <span class="k">def</span> <span class="nf">__repr__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Return a string representation of the chromosome including all the symbols in all the genes.</span>
<span class="sd">        Mainly for debugging purpose.</span>

<span class="sd">        :return: a string</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="s1">&#39;</span><span class="si">{}</span><span class="s1">[</span><span class="se">\n\t</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="p">)</span> <span class="o">+</span> <span class="s1">&#39;,</span><span class="se">\n\t</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">repr</span><span class="p">(</span><span class="n">g</span><span class="p">)</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="bp">self</span><span class="p">)</span> <span class="o">+</span> \
               <span class="s1">&#39;</span><span class="se">\n</span><span class="s1">], linker=</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">linker</span><span class="p">)</span></div>


</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../../../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../../../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
    <li><a href="https://github.com/ShuhuaGao/geppy"><img src="../../../_static/geppy-icon.png" style="height: 30px; vertical-align: middle; padding-right: 1em;" />Project Homepage</a> &#187;</li>
	<li><a href="../../../index.html">geppy 0.1.0a0 documentation</a> &#187; </li>

          <li class="nav-item nav-item-1"><a href="../../index.html" >Module code</a> &#187;</li> 
      </ul>
    </div>
<div class="footer">
    &copy; Copyright 2018, Shuhua Gao.
    <br />
    Built on None.
    <a href="../../../bugs.html">Found a bug</a>?
    <br />
    Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.7.6.
</div>
<script type="text/javascript">
  (function() {
     var ga = document.createElement('script');
         ga.src = ('https:' == document.location.protocol ?
                'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    ga.setAttribute('async', 'true');
    document.documentElement.firstChild.appendChild(ga);
  })();
</script>

  </body>
</html>